### PR TITLE
[GH-2699] Make expressions cache concurrent

### PR DESCRIPTION
**Pull request by Cameronjmayfield**

### CODE REVIEW

**Review by artembilan — CHANGES_REQUESTED** [review]
Please, add your name to the `@author` list and update Copyright over there to the current year.

Also would be great to have some test-case on the matter in the `MethodAnnotationPublisherMetadataSourceTests`.

Thank you for the contribution anyway!

**Comment by Cameronjmayfield** [comment]
@artembilan I'm not sure I can reliably reproduce the concurrent modification exception, so I added a test that just checks the instance of the map. Let me know what you think.

File: spring-integration-core/src/test/java/org/springframework/integration/aop/MethodAnnotationPublisherMetadataSourceTests.java [hunk 1]
```diff
@@ -83,6 +84,13 @@ public void payloadAndHeaders() {
 		assertEquals("#args[2]", headerMap.get("bar").getExpressionString());
 	}
 
+	@Test
+	public void expressionsAreConcurrentHashMap() {
+		assertTrue("Expressions should be concurrent to allow startup", getField(source, "channels") instanceof ConcurrentHashMap);
```
Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/aop/MethodAnnotationPublisherMetadataSourceTests.java]: There is `assertThat()` and `Matchers.instanceOf()` for better failure report if that.

File: spring-integration-core/src/test/java/org/springframework/integration/aop/MethodAnnotationPublisherMetadataSourceTests.java [hunk 2]
```diff
@@ -16,25 +16,26 @@
...
 import org.springframework.messaging.handler.annotation.Payload;
 
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.reflect.Method;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static org.junit.Assert.*;
+import static org.springframework.test.util.ReflectionTestUtils.getField;
```
Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/aop/MethodAnnotationPublisherMetadataSourceTests.java]: Please, no asterisk imports. No, static imports for not allowed methods, like this `getField()`.

Also, please, try to keep the import order we support.

To be sure that you are right, please, run Gradle command: `/gradlew clean :spring-integration-core:check`.

This way you'll get tests passed and also Checkstyle rules checked.

**Review by artembilan — CHANGES_REQUESTED** [review]
This is very close to the merge phase.
Let us know if you can fix all those concerns or we should do that for your on merge.

Thanks

File: spring-integration-core/src/test/java/org/springframework/integration/aop/MethodAnnotationPublisherMetadataSourceTests.java (hunk 1 above)
Comment (Cameronjmayfield) [reply spring-integration-core/src/test/java/org/springframework/integration/aop/MethodAnnotationPublisherMetadataSourceTests.java]: Great suggestion. Updated.

File: spring-integration-core/src/test/java/org/springframework/integration/aop/MethodAnnotationPublisherMetadataSourceTests.java (hunk 2 above)
Comment (Cameronjmayfield) [reply spring-integration-core/src/test/java/org/springframework/integration/aop/MethodAnnotationPublisherMetadataSourceTests.java]: Updated the style.

File: spring-integration-core/src/test/java/org/springframework/integration/aop/MethodAnnotationPublisherMetadataSourceTests.java [hunk 3]
```diff
@@ -83,6 +87,13 @@ public void payloadAndHeaders() {
 		assertEquals("#args[2]", headerMap.get("bar").getExpressionString());
 	}
 
+	@Test
+	public void expressionsAreConcurrentHashMap() {
+		assertThat("Expressions should be concurrent to allow startup", ReflectionTestUtils.getField(source, "channels"), instanceOf(ConcurrentHashMap.class));
+		assertThat("Expressions should be concurrent to allow startup", ReflectionTestUtils.getField(source, "payloadExpressions"), instanceOf(ConcurrentHashMap.class));
+		assertThat("Expressions should be concurrent to allow startup", ReflectionTestUtils.getField(source, "headersExpressions"), instanceOf(ConcurrentHashMap.class));
```
Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/aop/MethodAnnotationPublisherMetadataSourceTests.java]: One more code style concern and we will go for merge.
We prefer the code line length not more than `120` symbols.

I'm sharing all these rules with you to give your a chance to contribute more, but in much faster manner.
You won't see similar comments from me nay more afterwards! 😄

**Review by artembilan — CHANGES_REQUESTED** [review]

File: spring-integration-core/src/test/java/org/springframework/integration/aop/MethodAnnotationPublisherMetadataSourceTests.java (hunk 3 above)
Comment (Cameronjmayfield) [reply spring-integration-core/src/test/java/org/springframework/integration/aop/MethodAnnotationPublisherMetadataSourceTests.java]: I appreciate it. Updated.

**Review by artembilan — APPROVED** [review]

**Comment by artembilan** [comment]
... and cherry-picked to `5.0.x`.

@Cameronjmayfield ,

thank you very much for contribution and your patience.
Looking forward for more!

**Pull request merged**